src/hypertable_restrict_info.c

```diff
@@ -0,0 +1,171 @@
+#include "hypertable.h"
+
+static inline Dimension *
+get_dimension_for_partition_column_var(Var *var_expr, Hypertable *ht, Query *parse)
+{
+	RangeTblEntry *rte = rt_fetch(var_expr->varno, parse->rtable);
+	char	   *varname;
+
+	if (rte->relid != ht->main_table_relid)
+		return NULL;
+
+	varname = get_rte_attribute_name(rte, var_expr->varattno);
+
+	return hyperspace_get_dimension_by_name(ht->space, DIMENSION_TYPE_CLOSED, varname);
+}
+
+static Const *
+get_const_node(Node *node)
+{
+	if (IsA(node, Const))
+		return (Const *) node;
+	/* try to simplify the non-var expression */
+	node = eval_const_expressions(NULL, node);
+
+	if (IsA(node, Const))
+		return (Const *) node;
+
+	return NULL;
+}
+
+/* Return true if expr is var = const or const = var. If true, set
+ * var_node and const_node to the var and const. */
+static bool
+is_var_equal_const_opexpr(OpExpr *opexpr, Var **var_node, Const **const_node)
+{
+	Node	   *left;
+	Node	   *right;
+	Oid			eq_oid;
+
+	if (list_length(opexpr->args) != 2)
+		return false;
+
+
+	left = linitial(opexpr->args);
+	right = lsecond(opexpr->args);
+
+
+	if (IsA(left, Var))
+	{
+		*var_node = (Var *) left;
+		*const_node = get_const_node(right);
+	}
+	else if (IsA(right, Var))
+	{
+		*var_node = (Var *) right;
+		*const_node = get_const_node(left);
+	}
+	else
+	{
+		return false;
+	}
+
+	if (*const_node == NULL)
+		return false;
+
+
+	eq_oid = OpernameGetOprid(list_make2(makeString("pg_catalog"), makeString("=")), exprType(left), exprType(right));
+
+	if (eq_oid != opexpr->opno)
+		return false;
+
+	return true;
+}
+
+static HypertableRestrictInfo *
+hypertable_restrict_info_create(Dimension *dim, int32 hash_value)
+{
+	HypertableRestrictInfo *hre = palloc(sizeof(HypertableRestrictInfo));
+
+	*hre = (HypertableRestrictInfo)
+	{
+		.dim = dim,
+			.value = hash_value,
+	};
+	return hre;
+}
+
+
+HypertableRestrictInfo *
+hypertable_restrict_info_get(RestrictInfo *base_restrict_info, Query *query, Hypertable *hentry)
+{
+	Var		   *var_expr;
+	Const	   *const_expr;
+	Expr	   *node = base_restrict_info->clause;
+	OpExpr	   *opexpr;
+
+	if (!IsA(node, OpExpr))
+		return NULL;
+
+	opexpr = (OpExpr *) node;
+
+	if (is_var_equal_const_opexpr(opexpr, &var_expr, &const_expr))
+	{
+		/*
+		 * I now have a var = const. Make sure var is a partitioning column
+		 */
+		Dimension  *dim = get_dimension_for_partition_column_var(var_expr, hentry, query);
+
+		if (dim != NULL)
+			return hypertable_restrict_info_create(dim,
+												   partitioning_func_apply(dim->partitioning, const_expr->constvalue));
+	}
+	return NULL;
+}
+
+
+void
+hypertable_restrict_info_apply(HypertableRestrictInfo *spe,
+							   PlannerInfo *root,
+							   Hypertable *ht,
+							   Index main_table_child_index)
+{
+	ListCell   *lsib;
+
+	/*
+	 * Go through all the append rel list and find all children that are
+	 * chunks. Then check the chunk for exclusion and make dummy if necessary
+	 */
+	foreach(lsib, root->append_rel_list)
```

Comment:
As an alternative, would it be faster to just scan our metadata for chunks that are in the given dimension slice and then rebuild the append_rel_list from the result instead of scanning all chunks linearly?

Also, are time-dimension chunks excluded at this point or can we exclude them here in the same way?

Reply:
Btw, I played with the idea before about doing this at parsing time, by replacing the range table with a union of all the chunks. That would avoid a bunch of locking on chunks, for instance.

Reply:
Yup, I think we should do this for time-dimensions eventually. I think this takes quite a bit more plumbing to get right since we have to find all <, > comparisons etc. I've been watching this thread on hackers to see their solution: https://www.postgresql.org/message-id/flat/CAKJS1f-kPdMv9UyBkCeCi3bzXtOArGi-JS0zrF_fksyJj7Vp8w%40mail.gmail.com#CAKJS1f-kPdMv9UyBkCeCi3bzXtOArGi-JS0zrF_fksyJj7Vp8w@mail.gmail.com 

I think this is all doable but would rather do in a separate PR and probably after the -hackers stuff is settled. 

I am a bit wary of doing stuff in the parser because things are not yet expanded (i.e. views won't work) and I think you have a lot less context than in the planner. We may consider doing more surgery in the planner tho. One possibility is to set the `inh` flag on main tables to false before calling standard_planner so that main tables are not expanded and then doing the expansion ourselves.